client/tinychain/ml/dnn.py

```diff
@@ -12,35 +12,135 @@ def layer(weights, bias, activation):
+            ki = (1000,3)    1000 rows of data , 3 neurons this layer
+                             3,5  * 1000,3 -> 1000,5
+            kj   k = 1000 , j = 5
+            so basically einstein sum here is just simple np.dot(self.inputs,weights)
+            '''
+            
             return activation.forward(einsum("ij,ki->kj", [self.weights, inputs])) + self.bias
 
+
         def gradients(self, A_prev, dA, Z):
+            
+            '''
+            gradients function :it calculates delta values in Weights , Bias and in previous layer.
+            A_prev : Previous Input.
+            Z : Current Output before Activation.
+            dA: detla_Error.
+            dZ : delta Error in the Current Layer.
+            dA_Prev: delta Error in Previous Layer.
+            d_weights: delta Weights.
+            d_bias : delta Bias.
+            activation.backward(dA,Z) : performs derivative version of your activation function
+                                          to apply backward propagation.
+            
+            '''
             dZ = activation.backward(dA, Z).copy()
-            dA_prev = einsum("kj,ij->ki", [dZ, self[0]])
+            dA_prev = einsum("kj,ij->ki", [dZ, self.weights])  
             d_weights = einsum("kj,ki->ij", [dZ, A_prev])
             d_bias = dZ.sum(0)
             return dA_prev, d_weights, d_bias
 
         def train_eval(self, inputs):
+            '''
+            train_eval function you can use it on any layer to get its
+            output after dot product and after applying activation function on that output
+            
+            Z : Output of dot product between weight and inputs .
+            A : returning the output after applying activation function on it.
+            '''
             Z = einsum("ij,ki->kj", [self.weights, inputs])
             A = activation.forward(Z) + self.bias
             return A, Z
 
         def update(self, d_weights, d_bias):
+            '''
+            update function is used to ensure that update operations are done
+            New_Weight = Old Weight - Delta Weight
+            New_Bias = Old Bias - Delta Bias
+            '''
             return self.weights.write(self.weights - d_weights), self.bias.write(self.bias - d_bias)
+                
 
     return DNNLayer([weights, bias])
 
+    
+    """
+    
+    To Build Adam Optimizer or w/e you need just follow these steps:-
+    1. Create Your Class with Your Optimizer Name.
+    2. Build `Create` method as a class method that would initialize everything you need to set before optimizing
+    3. Build `Optimize` method to be called on each layer so it can return delta weights and delta bias also if 
+       you need to update some iterative parameters you might need to map all of them in function return.
+    
+    """
+    
+class Adam_Optimizer():
+    @classmethod
+    def create(cls):
+        return tc.Map({"l_m":tc.Number(0) ," l_v":tc.Number(0) , "t":tc.Number(0)})
```

Comment:
I think this should return an instance of type `cls`, meaning `AdamOptimizer` should be a subclass of `tc.Map` (or really a subclass of an abstract `Optimizer` class which is a subclass of `Map`) and then you can reference `l_m`, `l_v`, and `t` using the dot syntax (`optimizer_state.l_m`). Then the `optimize` method would need to return a new instance of `self.__class__` instead of `tc.Map`.

If you strongly prefer the way you have it, you should have `AdamOptimizer` subclass `object` and `create` can be a `staticmethod` instead of a `classmethod` (since you're not using `cls`).